### PR TITLE
Refactor for simplification and unique responsability for some classes Add OS support for macOS and Windows

### LINKED ISSUE #21902: Discrepancies in vulnerabilities reported for the same environment with 4.7.2 and 4.8.0
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| v4.8.0-beta1 | Vulnerability Detector | Manager/Agent | Packages | Ubuntu 22, Windows 11, RHEL 9, Amazon Linux 2023, macOS Sonoma |

## Description

During the testing performed on https://github.com/wazuh/wazuh-qa/issues/4898, it was found that for two identical environments (same clean box), the vulnerabilities reported by Vulnerability Detector in 4.7.2 and 4.8.0 are not the same:

OS|Box |4.7.2 Vulnerabilities|4.8.0 Vulnerabilities|
----| ---| ---- | ---| 
macOS Sonoma|macos_1400 (WazuhQAEnviroment)|[99](https://github.com/wazuh/wazuh-qa/files/14260796/macos-clean-vulns.json)| [2](https://github.com/wazuh/wazuh-qa/files/14239580/vulnerabilities_inventory-2024-02-12T14.35.09.298Z.csv) | 
Windows 11 | windows_11 (WazuhQAEnviroment)|[523](https://github.com/wazuh/wazuh-qa/files/14261249/win-clean-vulns.json)| 0 | 
Amazon Linux 2023 |gbailey/al2023| [0](https://github.com/wazuh/wazuh-qa/files/14223808/alas23-clean-vulns.json)|  [1500](https://github.com/wazuh/wazuh/files/14281621/vulnerabilities_inventory-2024-02-14T15.24.16.353Z.csv) |   
Ubuntu 22.04 |ubuntu/jammy64 | [178](https://github.com/wazuh/wazuh-qa/files/14223809/ubuntu-clean-vulns.json) | [34](https://github.com/wazuh/wazuh-qa/files/14238245/vulnerabilities_inventory-2024-02-12T12.17.00.564Z.csv) | 
RHEL 9 | AWS AMI |[62](https://github.com/wazuh/wazuh-qa/files/14278889/rhel-clean-vulns.json)| [2318](https://github.com/wazuh/wazuh-qa/files/14278892/vulnerabilities_inventory-2024-02-14T10.52.18.497Z.csv) |

### PR DESCRIPTION
**Pull request by Dwordcito**
Closes #21902 

### Description
This pr aims to add support for os scanning and code/classes simplification.

Basically based on the factory pattern and the chain of responsabilities pattern we can remove some conditions and establish a single responsability for each class.

### CODE REVIEW

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/alertClearBuilder.hpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+     * @brief Handles request and passes control to the next step of the chain.
+     *
+     * @param data Scan context.
+     * @return std::shared_ptr<ScanContext> Abstract handler.
+     */
+    std::shared_ptr<TScanContext> handleRequest(std::shared_ptr<TScanContext> data) override
+    {
+        // If the database is empty, we don't need to send the alert.
+        // If the database is empty, this means that the vulnerability scanner not have detected any package in
+        // the system. So, we don't need to send the alert.
```
Comment (GabrielEValenzuela) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/alertClearBuilder.hpp:62]:
```suggestion
     // An empty database indicates that the vulnerability scanner hasn't detected any packages in the system, thus no alert is necessary.
```

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/cleanAgentInventory.hpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                        elementKey.append("_");
+                        elementKey.append(cve);
+
+                        // We use elementKey as the id of the element, because it is unique.
+                        auto context = std::make_shared<TScanContext>();
+                        context->m_elements.emplace(elementKey,
+                                                    TInventorySync<TScanContext>::buildElement("DELETED", elementKey));
+                        m_subOrchestration->handleRequest(std::move(context));
+                    }
+                    logDebug2(WM_VULNSCAN_LOGTAG, "Deleting all agent vulnerabilities key: %s", dbQuery.first.c_str());
```
Comment (GabrielEValenzuela) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/cleanAgentInventory.hpp:99]: Should this log message be improved and contextualized depending on what type of element was deleted?

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/cleanInventory.hpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                    elementKey.append("_");
+                    elementKey.append(cve);
+
+                    // We use elementKey as the id of the element, because it is unique.
+                    auto context = std::make_shared<TScanContext>();
+                    context->m_elements.emplace(elementKey,
+                                                TInventorySync<TScanContext>::buildElement("DELETED", elementKey));
+                    m_subOrchestration->handleRequest(std::move(context));
+                }
+                logDebug2(WM_VULNSCAN_LOGTAG, "Deleting all agent vulnerabilities key: %s", dbQuery.first.c_str());
```
Comment (GabrielEValenzuela) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/cleanInventory.hpp:84]: Ditto

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/eventDeleteInventory.hpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
+/*
+ * Wazuh Vulnerability scanner - Scan Orchestrator
+ * Copyright (C) 2015, Wazuh Inc.
+ * Jun 1, 2023.
```
Comment (GabrielEValenzuela) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/eventDeleteInventory.hpp:4]:
```suggestion
 * March 4, 2024
 ```

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/eventDeleteInventory.hpp [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            for (const auto& cve : listCve)
+            {
+                std::string elementKey;
+                elementKey.append(key);
+                elementKey.append("_");
+                elementKey.append(cve);
+
+                data->m_elements.emplace(elementKey, TInventorySync<TScanContext>::buildElement("DELETED", elementKey));
+            }
+            logDebug2(WM_VULNSCAN_LOGTAG, "Deleting agent package key: %s", key.c_str());
```
Comment (GabrielEValenzuela) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/eventDeleteInventory.hpp:74]: Ditto

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/osScanner.hpp [hunk 6]
```diff
@@ -13,27 +13,294 @@
...
+        }
+
+        const auto osCPE = ScannerHelper::parseCPE(data->osCPEName().data());
+
+        auto vulnerabilityScan =
+            [&](const std::string& cnaName, const NSVulnerabilityScanner::ScanVulnerabilityCandidate& callbackData)
+        {
+            try
+            {
+                VersionObjectType objectType = VersionObjectType::DPKG;
```
Comment (GabrielEValenzuela) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/osScanner.hpp:76]: By default, we now assume that the probability of `VersionObjectType::DPKG` is greater, right?

**Review by GabrielEValenzuela — COMMENTED** [review]
Great job Octa!

Just a few comments and questions. I start with the testing phase!

**Comment by sebasfalcone** [comment]
# Testing
## Environment
- Manager: Linux mint 21.1 (v4.8.0)
- Agent: centOS 8 (v4.7.2)

## Configuration
```xml
  <wodle name="syscollector">
    <disabled>no</disabled>
    <interval>1h</interval>
    <scan_on_start>yes</scan_on_start>
    <hardware>no</hardware>
    <os>yes</os>
    <network>no</network>
    <packages>yes</packages>
    <ports all="no">no</ports>
    <processes>no</processes>

    <!-- Database synchronization settings -->
    <synchronization>
      <max_eps>10</max_eps>
    </synchronization>
  </wodle>
```

### Manager
#### First scan:

- Baseline
  - 49 vulnerabilities 

- Install firefox -> Expect an increase in vulnerabilities
	- 28 alerts generated 🟢 
	- 28 alerts indexed (active) 🟢 
	- 77 vulnerabilities 🟢 
- Remove firefox -> vulnerabilities decrease in vulnerabilities
	- 28 alerts generated 🟢 
	- 28 alerts indexed (solved) 🟢 
	- 49 vulnerabilities 🟢 

##3# Trigger re-scan

- Disable VD on manager -> Cleanup expected
	- 0 vulnerabilidades 🟢 

- Enable VD on manager -> Re-scan expected
	- 50 vulnerabilities 🟡 
	- vulnerabilities are not the same amount as in the first scan



#### Agent

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/packageScanner.hpp [hunk 7]
```diff
@@ -279,9 +283,7 @@ class TPackageScanner final : public AbstractHandler<std::shared_ptr<TScanContex
                               callbackData.cveId()->str().c_str());
 
                     data->m_elements[callbackData.cveId()->str()] = nlohmann::json::object();
-                    data->m_alerts[callbackData.cveId()->str()]["vulnerability"]["package"]["condition"] =
-                        "Package default status";
-
+                    data->m_matchConditions[callbackData.cveId()->str()] = {"", MatchRuleCondition::DefaultStatus};
                     return true;
                 }
```
Comment (MiguelazoDS) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/packageScanner.hpp:289]: This is not part of this implementation, but the Match by default status does not present any agent information as other match cases. 

```console
2024/03/08 16:31:52 wazuh-modulesd:vulnerability-scanner[55886] packageScanner.hpp:280 at operator()(): DEBUG: Match found for Package: policykit-1 for vulnerability: CVE-2016-2568 due to default status.
```

File: src/shared_modules/utils/rocksDBWrapper.hpp [hunk 8]
```diff
@@ -492,20 +492,91 @@ namespace Utils
...
+                    columnsNames.push_back((*it)->GetName());
+
+                    status = m_db->DestroyColumnFamilyHandle(*it);
+                    if (!status.ok())
+                    {
+                        std::cerr << "Failed to free RocksDB column family: " + std::string {status.getState()}
+                                  << std::endl;
+                    }
+
+                    it = m_columnsHandles.erase(it);
```
Comment (sebasfalcone) [inline src/shared_modules/utils/rocksDBWrapper.hpp:516]: Re-generate the column here to better atomize the method, if we fail to remove the following column we will still be in a working state
```suggestion
                    it = m_columnsHandles.erase(it);
                    createColumn(columnName);                    
```

File: src/shared_modules/utils/rocksDBWrapper.hpp [hunk 9]
```diff
@@ -492,20 +492,91 @@ namespace Utils
...
-                for (it->SeekToFirst(); it->Valid(); it->Next())
+                if ((*it)->GetName() != rocksdb::kDefaultColumnFamilyName)
                 {
-                    batch.Delete(it->key());
+                    auto status = m_db->DropColumnFamily(*it);
+                    if (!status.ok())
+                    {
+                        throw std::runtime_error("Error deleting data: " + status.ToString());
+                    }
+                    columnsNames.push_back((*it)->GetName());
```
[the remark this answers is not in the document]
Comment (sebasfalcone) [reply src/shared_modules/utils/rocksDBWrapper.hpp:507]: In concordance with my other comment:
```suggestion
                    const auto columnsName = (*it)->GetName();
```

File: src/shared_modules/utils/rocksDBWrapper.hpp [hunk 10]
```diff
@@ -492,20 +492,91 @@ namespace Utils
          */
         void deleteAll() override
         {
-            // Delete from all family columns
-            for (const auto& columnHandle : m_columnsHandles)
+            auto it = m_columnsHandles.begin();
+            std::vector<std::string> columnsNames;
```
Comment (sebasfalcone) [inline src/shared_modules/utils/rocksDBWrapper.hpp:496]: In concordance with my other comment:
```suggestion
```

**Comment by GabrielEValenzuela** [comment]
# Testing phase

During the last Fryday and today, a deep testing was performed in this PR.

The manager used was a `Ubuntu 22`, and the following steps was performed:

- Init a fresh scan: 🟢 
- Re scan: 🟢 
- Install a vulnerable package: 🟢 
- Remove a vulnerable package: 🟢 
Observations: A error related with #22440 was found, and if the syscollector time scan is too small, the scan show inconsistencies.

For agent:
RockyLinux8
- Init a fresh scan: 🟢 
- Remove agent: 🟢 
- Install a vulnerable package: 🟢 
- Remove a vulnerable package: 🟢 
Observations: Alerts related with that agent, are not erased.

Windows 10 - 22h2 
- Init a fresh scan: 🟡 
- Remove agent: 🟡  
- Install a vulnerable package: 🔴 
- Remove a vulnerable package: 🔴 
Observations: In the windows case, the vulnerabilities was few compared with the Friday test, the OS vulnerabilities and the package vulnerabilities. Also, each time a delta message was sent for the agent, the scan perform a re-scan of all. This could cause a overload if the syscollector time is small and the number of agents grow.

Windows Server 2022
- TBW

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/osScanner.hpp (hunk 6 above)
Comment (Dwordcito) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/osScanner.hpp:76]: yes it is correct.

**Review by GabrielEValenzuela — APPROVED** [review]
Amazing job Octa! 🚀

**Pull request merged**